core/variant/reduz_struct_proposal.h

```diff
@@ -0,0 +1,255 @@
+	_FORCE_INLINE_ static const StructMember& get_member(uint32_t p_index) {\
+	CRASH_BAD_INDEX(p_index,member_count)\
+	static StructMember members[member_count]={ __VA_ARGS__ };\
+	return members[p_index];\
+	}\
+};
+
+// Note GET_ARGUMENT_COUNT is a macro that we probably need to add tp typedefs.h, see:
+// https://stackoverflow.com/questions/2124339/c-preprocessor-va-args-number-of-arguments
+
+// Okay, so what is Struct<> ?
+
+// Its a similar class to TypedArray
+
+
+template <class T>
+class Struct : public Array {
+public:
+	typedef Type T;
+
+	_FORCE_INLINE_ void operator=(const Array &p_array) {
+	ERR_FAIL_COND_MSG(!is_same_typed(p_array), "Cannot assign a Struct from array with a different format.");
+	_ref(p_array);
+	}
+	_FORCE_INLINE_ Struct(const Variant &p_variant) :
+			Array(T::member_count, T::get_member,Array(p_variant)) {
+	}
+	_FORCE_INLINE_ Struct(const Array &p_array) :
+			Array(T::member_count, T::get_member,p_array) {
+	}
+	_FORCE_INLINE_ Struct() {
+		Array(T::member_count, T::get_member) {
+		}
+	};
+
+	// You likely saw correctly, we pass pointer to T::get_member. This is because we can't pass a structure and we want to initialize ArrayPrivate efficiently without allocating extra memory than needed, plus we want to keep this function around for validation:
+
+	Array::Array(uint32_t p_member_count, const StructMember& (*p_get_member)(uint32_t));
+	Array::Array(uint32_t p_member_count, const StructMember& (*p_get_member)(uint32_t),const Array &p_from); // separate one is best for performance since Array() does internal memory allocation when constructed.
+
+// Keep in mind also that GDScript VM is not able to pass a function pointer since this is dynamic, so it will need a separate constructor to initialize the array format. Same reason why the function pointer should not be kept inside of Array.
+// Likewise, GDScript may also need to pass a Script for class name, which is what ContainerTypeValidate neeeds.
+
+// Registering the struct to Class DB
+// call this inside _bind_methods of the relevant class
+
+// goes in object.h
+#define BIND_STRUCT(m_name) ClasDB::register_struct( m_name::get_class(), m_name::get_name(),  m_name::member_count, m_name::get_member);
+
+	Then you will also have to add this function `Array ClassDB::instantiate_struct(const StringName &p_class, const StringName& p_struct);` in order to construct them on demand.
+
+			// Optimizations:
+
+			// The idea here is that if GDScript code is typed, it should be able to access everything without any kind of validation or even copies. I will add this in the GDScript optimization proposal I have soon (pointer addressing mode).
+
+			// That said, I think we should consider changing ArrayPrivate::Array from Vector to LocalVector, this should enormously improve performance when accessing untyped (And eventually typed) arrays in GDScript. Arrays are shared, so there is not much of a need to use Vector<> here.
```

Comment:
I have wondered about that too, though I feel it is beyond the scope of this PR.